lib/ansible/modules/storage/netapp/na_ontap_qos_adaptive_policy_group.py

```diff
@@ -0,0 +1,335 @@
+        peak_iops_allocation: allocated_space
+        hostname: 10.193.78.30
+        username: admin
+        password: netapp1!
+
+    - name: modify adaptive qos policy group expected iops
+      na_ontap_qos_adaptive_policy_group:
+        state: present
+        name: aq_policy_1
+        vserver: policy_vserver
+        absolute_min_iops: 70IOPS
+        expected_iops: 125IOPS/TB
+        peak_iops: 250IOPS/TB
+        peak_iops_allocation: allocated_space
+        hostname: 10.193.78.30
+        username: admin
+        password: netapp1!
+
+    - name: modify adaptive qos policy group peak iops allocation
+      na_ontap_qos_adaptive_policy_group:
+        state: present
+        name: aq_policy_1
+        vserver: policy_vserver
+        absolute_min_iops: 70IOPS
+        expected_iops: 125IOPS/TB
+        peak_iops: 250IOPS/TB
+        peak_iops_allocation: used_space
+        hostname: 10.193.78.30
+        username: admin
+        password: netapp1!
+
+    - name: delete qos policy group
+      na_ontap_qos_adaptive_policy_group:
+        state: absent
+        name: aq_policy_1
+        vserver: policy_vserver
+        hostname: 10.193.78.30
+        username: admin
+        password: netapp1!
+
+"""
+
+RETURN = """
+"""
+
+import traceback
+
+import ansible.module_utils.netapp as netapp_utils
+from ansible.module_utils.netapp_module import NetAppModule
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+
+HAS_NETAPP_LIB = netapp_utils.has_netapp_lib()
+
+
+class NetAppOntapAdaptiveQosPolicyGroup(object):
+    """
+    Create, delete, modify and rename a policy group.
+    """
+    def __init__(self):
+        """
+        Initialize the Ontap qos policy group class.
+        """
+        self.argument_spec = netapp_utils.na_ontap_host_argument_spec()
+        self.argument_spec.update(dict(
+            state=dict(required=False, type='str', choices=['present', 'absent'], default='present'),
+            name=dict(required=True, type='str'),
+            from_name=dict(required=False, type='str'),
+            vserver=dict(required=True, type='str'),
+            absolute_min_iops=dict(required=False, type='str'),
+            expected_iops=dict(required=False, type='str'),
+            peak_iops=dict(required=False, type='str'),
+            peak_iops_allocation=dict(choices=['allocated_space', 'used_space'], default='used_space'),
+            force=dict(required=False, type='bool', default=False)
+        ))
+
+        self.module = AnsibleModule(
+            argument_spec=self.argument_spec,
+            supports_check_mode=True
+        )
+        self.na_helper = NetAppModule()
+        self.parameters = self.na_helper.set_parameters(self.module.params)
+
+        if HAS_NETAPP_LIB is False:
+            self.module.fail_json(
+                msg="the python NetApp-Lib module is required")
+        else:
+            self.server = netapp_utils.setup_na_ontap_zapi(
```

Comment:
If we are running this as a vserver admin (instead of cluster admin) the vserver needs to be passed into the setup function 
Here an example. 
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/storage/netapp/na_ontap_quotas.py#L173

Reply:
Looking at ZAPI, the qos-* API's are not available at the vserver level; it's only a cluster admin permitted operation at this point.

Reply:
The semantics in create are consistent with using a vserver in a cluster context, so I'm good with that.   But I don't see vserver being used in modify or delete.  Does it mean that a policy name is unique across a cluster ?   Or said otherwise, that 2 vservers cannot use the same policy name (or it points to the same policy).

Reply:
That's correct - policy names cannot conflict with the cluster-level default policies (extreme, performance, value). They must also be unique across the cluster; defining the same policy name within two vservers results in a ZAPI error.